pkg/dynamicinstrumentation/codegen/c/expressions.h

```diff
@@ -0,0 +1,309 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+#ifndef DI_EXPRESIONS_H
+#define DI_EXPRESIONS_H
+
+static __always_inline int read_register(struct expression_context context, __u64 reg, __u32 element_size)
+{
+    long err;
+    __u64 valueHolder = 0;
+    err = bpf_probe_read(&valueHolder,  element_size, &context.ctx->DWARF_REGISTER(reg));
+    if (err != 0) {
+        log_debug("error when reading data from register: %ld", err);
+    }
+    bpf_map_push_elem(&param_stack, &valueHolder, 0);
+    return err;
+}
+
+static __always_inline int read_stack(struct expression_context context, size_t stack_offset, __u32 element_size)
+{
+    long err;
+    __u64 valueHolder = 0;
+    err = bpf_probe_read(&valueHolder, element_size, &context.ctx->DWARF_STACK_REGISTER+stack_offset);
+    if (err != 0) {
+        log_debug("error when reading data from stack: %ld", err);
+    }
+    bpf_map_push_elem(&param_stack, &valueHolder, 0);
+    return err;
+}
+
+static __always_inline int read_register_value_to_output(struct expression_context context, __u64 reg, __u32 element_size)
+{
+    long err;
+    err = bpf_probe_read(&context.event->output[*(context.output_offset)], element_size, &context.ctx->DWARF_REGISTER(reg));
+    if (err != 0) {
+        log_debug("error when reading data while reading register value to output: %ld", err);
+    }
+    *context.output_offset += element_size;
+    return err;
+}
+
+static __always_inline int read_stack_value_to_output(struct expression_context context, __u64 stack_offset, __u32 element_size)
+{
+    long err;
+    err = bpf_probe_read(&context.event->output[*(context.output_offset)], element_size, &context.ctx->DWARF_STACK_REGISTER+stack_offset);
+    if (err != 0) {
+        log_debug("error when reading data while reading stack value to output: %ld", err);
+    }
+    *context.output_offset += element_size;
+    return err;
+}
+
+static __always_inline int pop(struct expression_context context, __u64 num_elements, __u32 element_size)
+{
+    long return_err;
+    long err;
+    __u64 valueHolder;
+    int i;
+    for(i = 0; i < num_elements; i++) {
+        bpf_map_pop_elem(&param_stack, &valueHolder);
+        log_debug("Popping to output: %llud", valueHolder);
+        err = bpf_probe_read(&context.event->output[*(context.output_offset)+i], element_size, &valueHolder);
+        if (err != 0) {
+            log_debug("error when reading data while popping from bpf stack: %ld", err);
+            return_err = err;
```

Comment:
Any reason you don't return early here?

Reply:
So we can attempt to pop the rest of the values, and at least move the buffer so it's moved to where we expect subsequent values.